### ISSUE TITLE
BPM container doesn't run successfully

### ISSUE THREAD

**Issue by WuMaKu** [AOT-Technologies/forms-flow-ai#35]
### Please provide information about your setup

- formsflow.ai Version: latest
- Browser: not related
- Browser version: not related

### Steps to Reproduce

1. Installed, configured & tested (using Postman) Keycloak Server
2. Neither the full deployment (tested on Docker for Windows & Docker on Ubuntu Server) nor
https://github.com/AOT-Technologies/forms-flow-ai/tree/master/forms-flow-bpm
brings up the BPM container (restarts all the time)

### Observed Behaviour
BPM Container crashes / restarts with error message:
2020-12-14 12:29:30.205 INFO 1 --- [ main] o.apache.http.impl.execchain.RetryExec : I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request to {}->http://localhost:8080: http protocol is not supported

2020-12-14 12:29:30.205 INFO 1 --- [ main] o.apache.http.impl.execchain.RetryExec : Retrying request to {}->http://localhost:8080

2020-12-14 12:29:30.206 INFO 1 --- [ main] o.apache.http.impl.execchain.RetryExec : I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request to {}->http://localhost:8080: http protocol is not supported

2020-12-14 12:29:30.206 INFO 1 --- [ main] o.apache.http.impl.execchain.RetryExec : Retrying request to {}->http://localhost:8080

2020-12-14 12:29:30.206 INFO 1 --- [ main] o.apache.http.impl.execchain.RetryExec : I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request to {}->http://localhost:8080: http protocol is not supported

2020-12-14 12:29:30.206 INFO 1 --- [ main] o.apache.http.impl.execchain.RetryExec : Retrying request to {}->http://localhost:8080

2020-12-14 12:29:30.207 ERROR 1 --- [ main] org.camunda.bpm.extension.keycloak : KEYCLOAK-01011 TOKEN request failed: I/O error on POST request for "http://localhost:8080/auth/realms/forms-flow-ai/protocol/openid-connect/token": http protocol is not supported; nested exception is org.apache.http.conn.UnsupportedSchemeException: http protocol is not supported
If possible provide a screenshot to demonstrate the bug.

### Expected Behaviour
The application should be up and available for use at port defaulted to 8000 in application.yaml http://localhost:8000/camunda/

**Comment by sumathi-thirumani-aot** [comment]
> 
> ## Bug Report
> ### Please provide information about your setup
> 
>     * formsflow.ai Version: latest
> 
>     * Browser: not related
> 
>     * Browser version: not related
> 
> 
> ### Steps to Reproduce
> 
>     1. Installed, configured & tested (using Postman) Keycloak Server
> 
>     2. Neither the full deployment (tested on Docker for Windows & Docker on Ubuntu Server) nor
>        https://github.com/AOT-Technologies/forms-flow-ai/tree/master/forms-flow-bpm
>        brings up the BPM container (restarts all the time)
> 
> 
> ### Observed Behaviour
> 
> BPM Container crashes / restarts with error message:
> 2020-12-14 12:29:30.205 INFO 1 --- [ main] o.apache.http.impl.execchain.RetryExec : I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request to {}->http://localhost:8080: http protocol is not supported
> 
> 2020-12-14 12:29:30.205 INFO 1 --- [ main] o.apache.http.impl.execchain.RetryExec : Retrying request to {}->http://localhost:8080
> 
> 2020-12-14 12:29:30.206 INFO 1 --- [ main] o.apache.http.impl.execchain.RetryExec : I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request to {}->http://localhost:8080: http protocol is not supported
> 
> 2020-12-14 12:29:30.206 INFO 1 --- [ main] o.apache.http.impl.execchain.RetryExec : Retrying request to {}->http://localhost:8080
> 
> 2020-12-14 12:29:30.206 INFO 1 --- [ main] o.apache.http.impl.execchain.RetryExec : I/O exception (org.apache.http.conn.UnsupportedSchemeException) caught when processing request to {}->http://localhost:8080: http protocol is not supported
> 
> 2020-12-14 12:29:30.206 INFO 1 --- [ main] o.apache.http.impl.execchain.RetryExec : Retrying request to {}->http://localhost:8080
> 
> 2020-12-14 12:29:30.207 ERROR 1 --- [ main] org.camunda.bpm.extension.keycloak : KEYCLOAK-01011 TOKEN request failed: I/O error on POST request for "http://localhost:8080/auth/realms/forms-flow-ai/protocol/openid-connect/token": http protocol is not supported; nested exception is org.apache.http.conn.UnsupportedSchemeException: http protocol is not supported
> If possible provide a screenshot to demonstrate the bug.
> ### Expected Behaviour
> 
> The application should be up and available for use at port defaulted to 8000 in application.yaml http://localhost:8000/camunda/

**Resolution:**
As log infers, http scheme is unsupported for integration with IDM.
Recommended solution would be to enable SSL for keycloak by following the instructions given on link https://www.keycloak.org/docs/latest/server_installation/
**Section to be referred : Enabling SSL/HTTPS for the Keycloak Server**

**Comment by WuMaKu** [comment]
Many thanks for your repsonse!
you might consider to set this as a requirement on this site:
https://github.com/AOT-Technologies/forms-flow-ai/blob/master/forms-flow-idm/keycloak-setup.md

I successfully started an https Keycloak Server (might also be useful to others:
sudo docker run --name keycloak -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=admin -p 8443:8443 jboss/keycloak

All the Postman requests worked but BPM gives me more errors... first error:
KEYCLOAK-01011 TOKEN request failed: I/O error on POST request for "https://localhost:8443/auth/realms/forms-flow-ai/protocol/openid-connect/token": Connect to localhost:8443 [localhost/127.0.0.1] failed: Connection refused (Connection refused); nested exception is org.apache.http.conn.HttpHostConnectException: Connect to localhost:8443 [localhost/127.0.0.1] failed: Connection refused (Connection refused)

I then changed the IP away from localhost, then I receive the following errors:
ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'springSecurityFilterChain' defined in class path resource [org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.servlet.Filter]: Factory method 'springSecurityFilterChain' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jwtDecoderByIssuerUri' defined in class path resource [org/springframework/boot/autoconfigure/security/oauth2/resource/servlet/OAuth2ResourceServerJwtConfiguration$JwtDecoderConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.oauth2.jwt.JwtDecoder]: Factory method 'jwtDecoderByIssuerUri' threw exception; nested exception is java.lang.IllegalArgumentException: Unable to resolve the Configuration with the provided Issuer of "https://172.22.104.155:8443/auth/realms/forms-flow-ai"

Caused by: org.springframework.web.client.ResourceAccessException: I/O error on GET request for "https://172.22.104.155:8443/auth/realms/forms-flow-ai/.well-known/openid-configuration": PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target; nested exception is javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

I assume that this is due to a self-signed certificate... is there any way to ignore "unsecure" certificates or how could a certificate be imported into bpm (as it is crashing / rebooting)?

Many thanks and best regards

**Comment by F5Brewery** [comment]
Agreed.  Please add this to the documentation it would help prevent hours/days of lost time.  
Following a bit of WuMaKu instruction I was also able to startup keycloak in SSL.  
I exported the cert for the site, imported into docker.  SEE https://stackoverflow.com/questions/35946844/docker-compose-pull-results-in-x509-certificate-signed-by-unknown-authority - With out this you get a different unauthorized error clearly pointed at keycloak.

I also had to go into the keycloak - client - forms-flow-bpm - settings - OpenID Connect Compatibility and turn on Use Refresh Tokens for Client Credentials Grant. Had a separate error about not being able to obtain token refresh.

Each step got me further...I then ended up here (seems to be same as WuMaKu):
Error creating bean with name 'springSecurityFilterChain' ......
Error creating bean with name 'jwtDecoderByIssuerUri' ......
Unable to resolve the Configuration with the provided Issuer of "https://[my.server.ip]:8443/auth/realms/forms-flow-ai"
PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target - Seems Java needs to also authorize cert...

Any assistance would be appreciated.

**Comment by F5Brewery** [comment]
I ended up bringing up a version of keycloak docker that used my SSL.  I was able to get this to get past the posted error by importing the cert into the java key store.  I did this by copying my cert into the forms-flow-bpm and editing the dockerfile before the WORKDIR command with:
RUN mkdir /certs
COPY rootcert.cer /certs/
RUN keytool -importcert -file /certs/rootcert.cer -cacerts -keypass changeit -storepass changeit -noprompt -alias myalias

If you let keycloak make its own ssl and you import that you end up with an issue about no SAN...blah blah.  I made my own SAN cert.

**Comment by F5Brewery** [comment]
I would be happy to write an addendum guide...
I now have all containers running with my self signed KeyCloak SSL instance.

in Redash I created a docker file where I did different things to try to get this to work.  In the end adding this environment variable line worked:
ENV REQUESTS_CA_BUNDLE /usr/local/share/ca-certificates/rootcert.crt

rootcert.crt is effectively my chained cert.  This could very well break the ability of Python to reach other external SSL sites- just a warning.

**Comment by WuMaKu** [comment]
Hi F5Brewery,
Happy to get some feedback :D! I'll try your instructions within the next few days, hopefully I can get it up and running too, many thanks!

**Comment by F5Brewery** [comment]
@WuMaKu 
You are quite welcome.
The export and import realm for keycloak saved me a ton of time and it works well.  Also do note that there is a missing setup for keycloak that is listed in a separate issue I have opened about missing tasks.
  
This is also some info I used for my keycloak:

https://stackoverflow.com/questions/52674979/keycloak-ssl-setup-using-docker-image
Bringing Keycloak up with a DB both in docker using docker-compose:
https://github.com/keycloak/keycloak-containers/tree/master/docker-compose-examples

I am more than happy to help in any way that I can.
